crates/ty_ide/src/completion.rs

```diff
@@ -1,40 +1,784 @@
+",
+        );
+
+        assert_snapshot!(test.completions(), @r"
+        foofoofoo
+        f
+        foofoo
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank1() {
+        let test = cursor_test(
+            "\
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+    <CURSOR>
+",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        //
+        // The tests below fail for the same reason that
+        // `nested_function_not_in_global_scope_blank` fails: there is no
+        // space in the AST ranges after the end of `foofoofoo` but before
+        // the end of `foofoo`. So either the AST needs to be tweaked to
+        // account for the indented whitespace, or some other technique
+        // needs to be used to get the scope containing `foofoo` but not
+        // `foofoofoo`.
+        assert_snapshot!(test.completions(), @r"
+        foo
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank2() {
+        let test = cursor_test(
+            " \
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+    <CURSOR>",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank3() {
+        let test = cursor_test(
+            "\
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+    <CURSOR>
+def frob(): ...
+            ",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank4() {
+        let test = cursor_test(
+            "\
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+    <CURSOR>
+
+def frob(): ...
+",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn double_nested_function_not_in_global_scope_blank5() {
+        let test = cursor_test(
+            "\
+def foo():
+    def foofoo():
+        def foofoofoo(): ...
+
+    <CURSOR>
+
+def frob(): ...
+",
+        );
+
+        // FIXME: Should include `foofoo` (but not `foofoofoo`).
+        assert_snapshot!(test.completions(), @r"
+        foo
+        frob
+        ");
+    }
+
+    #[test]
+    fn list_comprehension1() {
+        let test = cursor_test(
+            "\
+[<CURSOR> for bar in [1, 2, 3]]
+",
+        );
+
+        // It's not totally clear why `for` shows up in the
+        // symbol tables of the detected scopes here. My guess
+        // is that there's perhaps some sub-optimal behavior
+        // here because the list comprehension as written is not
+        // valid.
```

Comment:
This is due to error recovery. The source code contains a syntax error because the identifier is absent (`[ for bar in [1, 2, 3]`) and the parser recovers by converting `for` into an identifier. The fix is to improve error recovery where the parser is converting keywords into an identifier. There are benefits for doing this but maybe not doing this conversion might be more beneficial. Point 12 in #10653.